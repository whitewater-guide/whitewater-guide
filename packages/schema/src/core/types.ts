import { PageInfo } from '../__generated__/types';

/**
 * Relay-styles edge
 **/
export interface Edge<TNode> {
  node: TNode;
  cursor: CodegenCursor;
}

/**
 * Relay-styles connection
 **/
export interface RelayConnection<TNode> {
  edges: Array<Edge<TNode>>;
  pageInfo: PageInfo;
}
