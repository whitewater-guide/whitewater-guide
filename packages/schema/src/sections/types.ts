export enum Duration {
  LAPS = 10,
  TWICE = 20,
  DAYRUN = 30,
  OVERNIGHTER = 40,
  MULTIDAY = 50,
}

export const Durations = new Map<number, string>();
Durations.set(Duration.LAPS, 'laps');
Durations.set(Duration.TWICE, 'twice');
Durations.set(Duration.DAYRUN, 'full day');
Durations.set(Duration.OVERNIGHTER, 'overnighter');
Durations.set(Duration.MULTIDAY, 'multiday');

Durations.entries();
