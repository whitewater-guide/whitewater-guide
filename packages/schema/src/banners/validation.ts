import { yupSchemas } from '@whitewater-guide/validation';
import type { ObjectSchema } from 'yup';
import { array, bool, mixed, number, object, string } from 'yup';

import type { BannerInput, BannerSourceInput } from '../__generated__/types';
import { BannerKind, BannerPlacement } from '../__generated__/types';

const BannerSourceSchema: ObjectSchema<BannerSourceInput> = object({
  kind: mixed<BannerKind>().oneOf(Object.values(BannerKind)).required(),
  url: string().required(),
}).noUnknown();

export const BannerInputSchema: ObjectSchema<BannerInput> = object({
  id: string().uuid().nullable(),
  slug: string().slug().required(),
  name: string().required(),
  priority: number().integer().required(),
  enabled: bool().required(),
  placement: mixed<BannerPlacement>()
    .oneOf(Object.values(BannerPlacement))
    .required(),
  source: BannerSourceSchema.clone(),
  link: string().url().https().nullable(),
  extras: object().nullable(),
  regions: array(yupSchemas.refInput().required()).required(),
  groups: array(yupSchemas.refInput().required()).required(),
})
  .strict(true)
  .noUnknown(true);
