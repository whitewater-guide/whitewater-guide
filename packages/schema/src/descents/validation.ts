import type { ObjectSchema } from 'yup';
import { bool, date, number, object, string } from 'yup';

import type { DescentInput, DescentLevelInput } from '../__generated__/types';

export const DescentLevelInputSchema: ObjectSchema<DescentLevelInput> = object({
  unit: string().max(20).optional().nullable(),
  value: number().required(),
})
  .strict(true)
  .noUnknown();

export const DescentInputSchema: ObjectSchema<DescentInput> = object({
  id: string().uuid().optional().nullable(),
  sectionId: string().uuid().required().nonNullable(),
  startedAt: date().required(),
  duration: number().integer().min(0).notRequired().nullable(),
  level: DescentLevelInputSchema.clone().notRequired().nullable(),
  comment: string().notRequired().nullable(),
  public: bool().notRequired().nullable(),
})
  .strict(true)
  .noUnknown();
