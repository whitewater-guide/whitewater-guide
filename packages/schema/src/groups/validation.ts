import type { ObjectSchema } from 'yup';
import { object, string } from 'yup';

import type { GroupInput } from '../__generated__/types';

const GROUP_SKU = /^group\.\w{3,}$/;

export const GroupInputSchema: ObjectSchema<GroupInput> = object({
  id: string().uuid().defined().nullable(),
  name: string().nonEmpty(),
  sku: string()
    .defined()
    .max(255)
    .matches(GROUP_SKU, 'yup:string.sku')
    .nullable(),
})
  .strict(true)
  .noUnknown();
