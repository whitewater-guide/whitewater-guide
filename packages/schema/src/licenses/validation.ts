import type { ObjectSchema } from 'yup';
import { object, string } from 'yup';

import type { LicenseInput } from '../__generated__/types';

export const LicenseInputSchema: ObjectSchema<LicenseInput> = object({
  url: string().url().nullable(),
  name: string().min(2).defined().nonNullable(),
  slug: string().slug().nullable(),
})
  .strict(true)
  .noUnknown();
