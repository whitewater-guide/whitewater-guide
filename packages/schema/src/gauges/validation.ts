import { yupSchemas } from '@whitewater-guide/validation';
import type { ObjectSchema } from 'yup';
import { mixed, object, string } from 'yup';

import type { GaugeInput } from '../__generated__/types';
import { PointInputSchema } from '../points';

export const GaugeInputSchema: ObjectSchema<GaugeInput> = object({
  id: string().uuid().nullable(),
  name: string().nonEmpty(),
  code: string().nonEmpty(),
  levelUnit: string().nullable(),
  flowUnit: string().nullable(),
  timezone: string().nullable(),
  location: PointInputSchema.clone().nullable(),
  requestParams: mixed().nullable(),
  url: string().url().nullable(),
  source: yupSchemas.refInput().required(),
})
  .strict(true)
  .noUnknown();
