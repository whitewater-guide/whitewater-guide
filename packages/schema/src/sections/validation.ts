import { yupSchemas } from '@whitewater-guide/validation';
import times from 'lodash/times';
import type { ObjectSchema } from 'yup';
import { array, bool, mixed, number, object, string } from 'yup';

import type { GaugeBindingInput, SectionInput } from '../__generated__/types';
import { LicenseInputSchema } from '../licenses';
import { MediaInputSchema } from '../media';
import { CoordinateSchema, PointInputSchema } from '../points';
import { Durations } from './types';

export const GaugeBindingSchema: ObjectSchema<GaugeBindingInput> = object({
  __typename: mixed().optional(),
  minimum: number().nullable(),
  maximum: number().nullable(),
  optimum: number().nullable(),
  impossible: number().nullable(),
  approximate: bool().nullable(),
  formula: string().formula().nullable(),
})
  .strict(true)
  .noUnknown();

const SimpleTagSchema = object({
  id: string().required(),
  name: string().optional(),
})
  .strict(true)
  .noUnknown();

export const SectionInputSchema: ObjectSchema<SectionInput> = object({
  id: string().uuid().nullable(),
  name: string().nonEmpty(),
  altNames: array().of(string().required()).nullable(),
  description: string().nullable(),
  season: string().nullable(),
  seasonNumeric: array()
    .of(number().integer().defined().min(0).max(23))
    .max(24)
    .defined(),

  river: yupSchemas.newNode().required(),
  gauge: yupSchemas.refInput().optional().nullable(),
  region: yupSchemas.refInput().optional().nullable(),
  levels: GaugeBindingSchema.clone().nullable(),
  flows: GaugeBindingSchema.clone().nullable(),
  flowsText: string().nullable(),
  timezone: string().nullable(),

  shape: array().of(CoordinateSchema).min(2).required(),
  distance: number().positive().nullable(),
  drop: number().positive().nullable(),
  duration: number().oneOf(Array.from(Durations.keys())).nullable(),
  difficulty: number()
    .oneOf(times(13, (i) => i * 0.5))
    .required(),
  difficultyXtra: string().max(32).nullable(),
  rating: number()
    .oneOf(times(11, (i) => i * 0.5))
    .nullable(),
  tags: array().of(SimpleTagSchema.clone()).required(),
  pois: array().of(PointInputSchema.clone().required()).required(),
  media: array().of(MediaInputSchema.clone().required()).required(),
  hidden: bool().required(),
  helpNeeded: string().nullable(),
  createdBy: string().uuid().notRequired().nullable(),
  importId: string().nullable(),
  copyright: string().nullable(),
  license: LicenseInputSchema.clone().nullable(),
})
  .strict(true)
  .noUnknown();

export const SectionAdminSettingsSchema = object({
  demo: bool().defined(),
})
  .strict(true)
  .noUnknown();
