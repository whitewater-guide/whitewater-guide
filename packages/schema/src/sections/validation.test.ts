import { createSafeValidator } from '@whitewater-guide/validation';

import {
  GaugeBindingInput,
  MediaKind,
  SectionAdminSettings,
  SectionInput,
} from '../__generated__/types';
import { Duration } from './types';
import {
  GaugeBindingSchema,
  SectionAdminSettingsSchema,
  SectionInputSchema,
} from './validation';

describe('GaugeBinding', () => {
  const validator = createSafeValidator(GaugeBindingSchema);
  type TestValue = [string, GaugeBindingInput];
  type IncorrectTestValue = [string, any];

  const correct: GaugeBindingInput = {
    minimum: -10,
    maximum: 10,
    optimum: 10,
    impossible: 10,
    approximate: true,
    formula: 'x + 10',
  };

  const correctValues: TestValue[] = [
    ['full value', correct],
    [
      'null value',
      {
        minimum: null,
        maximum: null,
        optimum: null,
        impossible: null,
        approximate: null,
        formula: null,
      },
    ],
    ['empty value', {}],
  ];

  const incorrectValues: IncorrectTestValue[] = [
    ['incorrect formula', { ...correct, formula: 'x *' }],
    ['too many variables', { ...correct, formula: 'x * y + 2' }],
    ['bad variable', { ...correct, formula: 'X + 10' }],
    ['extra fields', { ...correct, foo: 'X + 10' }],
  ];

  it.each(correctValues)('should be valid for %s', (_, value) => {
    expect(validator(value)).toBeNull();
  });

  it.each(incorrectValues)('should be invalid for %s', (_, value) => {
    const error = validator(value);
    expect(error).not.toBeNull();
    expect(error).toMatchSnapshot();
  });
});

describe('SectionInput', () => {
  const validator = createSafeValidator(SectionInputSchema);
  type TestValue = [string, SectionInput];
  type IncorrectTestValue = [string, any];

  const required: SectionInput = {
    name: 'Section',
    difficulty: 3.5,
    river: { id: '1aa9fd9a-c34b-11e8-a355-529269fb1459' },
    seasonNumeric: [],
    shape: [
      [0, 0, 0],
      [0.1, 1, 0],
    ],
    hidden: false,
    tags: [],
    pois: [],
    media: [],
  };

  const correct: SectionInput = {
    ...required,
    id: 'eda6bd7e-c34a-11e8-a355-529269fb1459',
    altNames: ['Alt name 1', 'Alt name 2'],
    description: 'Section description',
    season: 'When it rains',
    seasonNumeric: [10, 11, 12],
    gauge: { id: '2bedcd8e-c34b-11e8-a355-529269fb1459' },
    levels: { minimum: 10, optimum: null },
    flows: { maximum: 11, approximate: true },
    flowsText: 'unknown',
    distance: 10,
    drop: 100,
    duration: Duration.LAPS,
    difficultyXtra: 'X',
    rating: 2.5,
    tags: [{ id: '4x4' }, { id: 'portage' }],
    pois: [
      {
        id: 'b5678064-c34b-11e8-a355-529269fb1459',
        coordinates: [1, 1.3, 200],
        description: 'poi description',
        name: 'some poi',
        kind: 'portage',
      },
    ],
    media: [
      {
        id: '9f0ff4b6-c258-11e8-a355-529269fb1459',
        description: 'description',
        kind: MediaKind.Photo,
        resolution: [800, 600],
        url: 'jhakjfh',
        weight: -3,
        copyright: 'Copyright',
        license: {
          slug: 'CC_BY-SA',
          name: 'Attribution-ShareAlike 4.0 International (CC BY-SA 4.0)',
          url: 'https://creativecommons.org/licenses/by-sa/4.0/',
        },
      },
    ],
    helpNeeded: 'fix duration',
    copyright: 'section copyright',
    license: {
      slug: 'CC_BY-SA',
      name: 'Attribution-ShareAlike 4.0 International (CC BY-SA 4.0)',
      url: 'https://creativecommons.org/licenses/by-sa/4.0/',
    },
    timezone: 'Europe/Moscow',
  };

  const { license: _l, copyright: _c, ...legacy } = correct;

  const correctValues: TestValue[] = [
    ['full value', correct],
    ['required fields only', required],
    [
      'null value',
      {
        ...correct,
        id: null,
        altNames: [],
        description: null,
        difficultyXtra: null,
        distance: null,
        drop: null,
        duration: null,
        flows: null,
        flowsText: null,
        gauge: null,
        levels: null,
        rating: null,
        season: null,
        createdBy: null,
        helpNeeded: null,
        importId: null,
        copyright: null,
        license: null,
      },
    ],
    [
      'undefined value',
      {
        ...correct,
        altNames: [],
        description: undefined,
        difficultyXtra: undefined,
        distance: undefined,
        drop: undefined,
        duration: undefined,
        flows: undefined,
        flowsText: undefined,
        gauge: undefined,
        levels: undefined,
        rating: undefined,
        season: undefined,
        createdBy: undefined,
        helpNeeded: undefined,
        importId: undefined,
      },
    ],
    [
      'with createdBy',
      { ...correct, createdBy: '880e1a40-bd02-11e9-9cb5-2a2ae2dbcce4' },
    ],
    ['with importId', { ...correct, importId: 'whatever' }],
    ['empty description', { ...correct, description: '' }],
    ['empty season', { ...correct, season: '' }],
    ['empty season numeric', { ...correct, seasonNumeric: [] }],
    ['empty flows text', { ...correct, flowsText: '' }],
    ['empty difficultyXtra', { ...correct, difficultyXtra: '' }],
    [
      'new river',
      {
        ...correct,
        river: { id: '__NEW_RIVER_ID__', name: 'New river' },
        region: { id: 'cf8c20ac-ae06-11e9-a2a3-2a2ae2dbcce4' },
      },
    ],
    ['empty media', { ...correct, media: [] }],
    ['legacy inpit without copyright and license', legacy],
  ];

  const incorrectValues: IncorrectTestValue[] = [
    ['bad id', { ...correct, id: 'foo' }],
    ['bad name', { ...correct, name: '' }],
    ['bad alt name', { ...correct, altNames: ['foo', '', 'bar'] }],
    [
      'bad season numeric (bad half-month)',
      { ...correct, seasonNumeric: [10, 100, -10, 11] },
    ],
    [
      'bad season numeric (too many items)',
      { ...correct, seasonNumeric: Array(30).fill(10) },
    ],
    ['bad river id', { ...correct, river: { id: 'foo' } }],
    ['bad gauge id', { ...correct, gauge: { id: 'foo' } }],
    ['bad shape (not enough points)', { ...correct, shape: [[1, 2, 3]] }],
    [
      'bad shape (bad point)',
      {
        ...correct,
        shape: [
          [1, 2, 3],
          [300, 1, 1],
          [2, 3, 4],
        ],
      },
    ],
    ['bad distance', { ...correct, distance: -100 }],
    ['bad drop', { ...correct, drop: -100 }],
    ['bad duration', { ...correct, duration: -100 }],
    ['bad difficulty (min)', { ...correct, difficulty: -0.5 }],
    ['bad difficulty (max)', { ...correct, difficulty: 7.5 }],
    ['bad difficulty (fraction)', { ...correct, difficulty: 4.4 }],
    ['bad extra difficulty', { ...correct, difficultyXtra: 'x'.repeat(50) }],
    ['bad rating (min)', { ...correct, rating: -0.5 }],
    ['bad rating (max)', { ...correct, rating: 5.5 }],
    ['bad rating (fraction)', { ...correct, rating: 4.4 }],
    ['bad tags', { ...correct, tags: ['foo', '', 'bar'] }],
    [
      'bad pois',
      {
        ...correct,
        pois: [
          {
            id: null,
            description: null,
            kind: 'dam',
            name: null,
            coordinates: [1, 1, 1],
          },
          {
            id: 'foo',
            description: null,
            kind: 'dam',
            name: null,
            coordinates: [1, 1, 1],
          },
          {
            id: null,
            description: null,
            kind: 'dam',
            name: null,
            coordinates: [1, 2, 1],
          },
        ],
      },
    ],
    ['null media', { ...correct, media: null }],
    [
      'incorrect media',
      { ...correct, media: [{ ...correct.media[0], id: 'foo' }] },
    ],
    ['extra fields', { ...correct, foo: 'bar' }],
    [
      'new river without name',
      { ...correct, river: { id: '__NEW_RIVER_ID__' } },
    ],
  ];

  it.each(correctValues)('should be valid for %s', (_, value) => {
    expect(validator(value)).toBeNull();
  });

  it.each(incorrectValues)('should be invalid for %s', (_, value) => {
    const error = validator(value);
    expect(error).not.toBeNull();
    expect(error).toMatchSnapshot();
  });
});

describe('SectionAdminSettings', () => {
  const validator = createSafeValidator(SectionAdminSettingsSchema);
  type TestValue = [string, SectionAdminSettings];
  type IncorrectTestValue = [string, any];

  const correct: SectionAdminSettings = { demo: true };

  const correctValues: TestValue[] = [['full value', correct]];
  const incorrectValues: IncorrectTestValue[] = [
    ['demo undefined', {}],
    ['extra fields', { ...correct, foo: 'bar' }],
  ];

  it.each(correctValues)('should be valid for %s', (_, value) => {
    expect(validator(value)).toBeNull();
  });

  it.each(incorrectValues)('should be invalid for %s', (_, value) => {
    const error = validator(value);
    expect(error).not.toBeNull();
    expect(error).toMatchSnapshot();
  });
});
