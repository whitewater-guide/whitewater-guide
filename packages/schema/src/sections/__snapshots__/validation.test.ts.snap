// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GaugeBinding should be invalid for bad variable 1`] = `
{
  "formula": "yup:string.formula",
}
`;

exports[`GaugeBinding should be invalid for extra fields 1`] = `
{
  "": {
    "key": "yup:object.noUnknown",
    "options": {
      "path": "this",
    },
  },
}
`;

exports[`GaugeBinding should be invalid for incorrect formula 1`] = `
{
  "formula": "yup:string.formula",
}
`;

exports[`GaugeBinding should be invalid for too many variables 1`] = `
{
  "formula": "yup:string.formula",
}
`;

exports[`SectionAdminSettings should be invalid for demo undefined 1`] = `
{
  "demo": "yup:mixed.defined",
}
`;

exports[`SectionAdminSettings should be invalid for extra fields 1`] = `
{
  "": {
    "key": "yup:object.noUnknown",
    "options": {
      "path": "this",
    },
  },
}
`;

exports[`SectionInput should be invalid for bad alt name 1`] = `
{
  "altNames": [
    ,
    "yup:mixed.required",
  ],
}
`;

exports[`SectionInput should be invalid for bad difficulty (fraction) 1`] = `
{
  "difficulty": {
    "key": "yup:mixed.oneOf",
    "options": {
      "values": "0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5, 5.5, 6",
    },
  },
}
`;

exports[`SectionInput should be invalid for bad difficulty (max) 1`] = `
{
  "difficulty": {
    "key": "yup:mixed.oneOf",
    "options": {
      "values": "0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5, 5.5, 6",
    },
  },
}
`;

exports[`SectionInput should be invalid for bad difficulty (min) 1`] = `
{
  "difficulty": {
    "key": "yup:mixed.oneOf",
    "options": {
      "values": "0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5, 5.5, 6",
    },
  },
}
`;

exports[`SectionInput should be invalid for bad distance 1`] = `
{
  "distance": {
    "key": "yup:number.positive",
    "options": {
      "path": "distance",
    },
  },
}
`;

exports[`SectionInput should be invalid for bad drop 1`] = `
{
  "drop": {
    "key": "yup:number.positive",
    "options": {
      "path": "drop",
    },
  },
}
`;

exports[`SectionInput should be invalid for bad duration 1`] = `
{
  "duration": {
    "key": "yup:mixed.oneOf",
    "options": {
      "values": "10, 20, 30, 40, 50",
    },
  },
}
`;

exports[`SectionInput should be invalid for bad extra difficulty 1`] = `
{
  "difficultyXtra": {
    "key": "yup:string.max",
    "options": {
      "max": 32,
      "path": "difficultyXtra",
    },
  },
}
`;

exports[`SectionInput should be invalid for bad gauge id 1`] = `
{
  "gauge": {
    "id": {
      "key": "yup:string.uuid",
      "options": {
        "path": "gauge.id",
      },
    },
  },
}
`;

exports[`SectionInput should be invalid for bad id 1`] = `
{
  "id": {
    "key": "yup:string.uuid",
    "options": {
      "path": "id",
    },
  },
}
`;

exports[`SectionInput should be invalid for bad name 1`] = `
{
  "name": "yup:string.nonEmpty",
}
`;

exports[`SectionInput should be invalid for bad pois 1`] = `
{
  "pois": [
    ,
    {
      "id": {
        "key": "yup:string.uuid",
        "options": {
          "path": "pois[1].id",
        },
      },
    },
  ],
}
`;

exports[`SectionInput should be invalid for bad rating (fraction) 1`] = `
{
  "rating": {
    "key": "yup:mixed.oneOf",
    "options": {
      "values": "0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5",
    },
  },
}
`;

exports[`SectionInput should be invalid for bad rating (max) 1`] = `
{
  "rating": {
    "key": "yup:mixed.oneOf",
    "options": {
      "values": "0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5",
    },
  },
}
`;

exports[`SectionInput should be invalid for bad rating (min) 1`] = `
{
  "rating": {
    "key": "yup:mixed.oneOf",
    "options": {
      "values": "0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5",
    },
  },
}
`;

exports[`SectionInput should be invalid for bad river id 1`] = `
{
  "river": "yup:mixed.newNode",
}
`;

exports[`SectionInput should be invalid for bad season numeric (bad half-month) 1`] = `
{
  "seasonNumeric": [
    ,
    {
      "key": "yup:number.max",
      "options": {
        "max": 23,
        "path": "seasonNumeric[1]",
      },
    },
    {
      "key": "yup:number.min",
      "options": {
        "min": 0,
        "path": "seasonNumeric[2]",
      },
    },
  ],
}
`;

exports[`SectionInput should be invalid for bad season numeric (too many items) 1`] = `
{
  "seasonNumeric": {
    "key": "yup:array.max",
    "options": {
      "max": 24,
      "path": "seasonNumeric",
    },
  },
}
`;

exports[`SectionInput should be invalid for bad shape (bad point) 1`] = `
{
  "shape": [
    ,
    [
      "yup:number.longitude",
    ],
  ],
}
`;

exports[`SectionInput should be invalid for bad shape (not enough points) 1`] = `
{
  "shape": {
    "key": "yup:array.min",
    "options": {
      "min": 2,
      "path": "shape",
    },
  },
}
`;

exports[`SectionInput should be invalid for bad tags 1`] = `
{
  "tags": [
    {
      "key": "yup:mixed.notType",
      "options": {
        "type": "object",
      },
    },
    {
      "key": "yup:mixed.notType",
      "options": {
        "type": "object",
      },
    },
    {
      "key": "yup:mixed.notType",
      "options": {
        "type": "object",
      },
    },
  ],
}
`;

exports[`SectionInput should be invalid for extra fields 1`] = `
{
  "": {
    "key": "yup:object.noUnknown",
    "options": {
      "path": "this",
    },
  },
}
`;

exports[`SectionInput should be invalid for incorrect media 1`] = `
{
  "media": [
    {
      "id": {
        "key": "yup:string.uuid",
        "options": {
          "path": "media[0].id",
        },
      },
    },
  ],
}
`;

exports[`SectionInput should be invalid for new river without name 1`] = `
{
  "river": "yup:mixed.newNode",
}
`;

exports[`SectionInput should be invalid for null media 1`] = `
{
  "media": "yup:mixed.required",
}
`;
